src/hotspot/share/gc/shared/freeListAllocator.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "gc/shared/freeListAllocator.hpp"
+#include "logging/log.hpp"
+#include "utilities/globalCounter.inline.hpp"
+
+FreeListAllocator::NodeList::NodeList() :
+  _head(NULL), _tail(NULL), _entry_count(0) {}
+
+FreeListAllocator::NodeList::NodeList(FreeNode* head, FreeNode* tail, size_t entry_count) :
+  _head(head), _tail(tail), _entry_count(entry_count)
+{
+  assert((_head == NULL) == (_tail == NULL), "invariant");
+  assert((_head == NULL) == (_entry_count == 0), "invariant");
+}
+
+FreeListAllocator::PendingList::PendingList() :
+  _tail(nullptr), _head(nullptr), _count(0) {}
+
+size_t FreeListAllocator::PendingList::add(FreeNode* node) {
+  assert(node->next() == nullptr, "precondition");
+  FreeNode* old_head = Atomic::xchg(&_head, node);
+  if (old_head != nullptr) {
+    node->set_next(old_head);
+  } else {
+    assert(_tail == nullptr, "invariant");
+    _tail = node;
+  }
+  return Atomic::add(&_count, size_t(1));
+}
+
+typename FreeListAllocator::NodeList FreeListAllocator::PendingList::take_all() {
+  NodeList result{Atomic::load(&_head), _tail, Atomic::load(&_count)};
+  Atomic::store(&_head, (FreeNode*)nullptr);
+  _tail = nullptr;
+  Atomic::store(&_count, size_t(0));
+  return result;
+}
+
+size_t FreeListAllocator::PendingList::count() const {
+  return  Atomic::load(&_count);
+}
+
+FreeListAllocator::FreeListAllocator(const char* name, FreeListConfig* config) :
+  _config(config),
+  _free_count(0),
+  _free_list(),
+  _transfer_lock(false),
+  _active_pending_list(0),
+  _pending_lists()
+{
+  strncpy(_name, name, sizeof(_name) - 1);
+  _name[sizeof(_name) - 1] = '\0';
+}
+
+void FreeListAllocator::delete_list(FreeNode* list) {
+  while (list != NULL) {
+    FreeNode* next = list->next();
+    DEBUG_ONLY(list->set_next(NULL);)
```

Comment:
Setting next to null isn't needed unless there's a null check in the destructor.  The current destructor is defaulted.  I think some predecessors to this code might have had a debug-only assert that next was null.